src/vdaf/hits.rs

```diff
@@ -0,0 +1,741 @@
+
+    // Derive the verification randomness.
+    let mut deriver = KeyDeriver::from_key(&verify_param.verify_rand_init);
+    deriver.update(nonce);
+    let verify_rand_seed = deriver.finish();
+    let mut verify_rand_prng: Prng<I::Field> =
+        Prng::from_key_stream(KeyStream::from_key(&verify_rand_seed));
+
+    // Evaluate the IDPF shares and compute the polynomial coefficients.
+    let mut z = [I::Field::zero(); 3];
+    let mut output_share = Vec::with_capacity(agg_param.prefixes.len());
+    for prefix in agg_param.prefixes.iter() {
+        let v = input_share.data.eval(prefix)?[0];
+        let k = input_share.auth.eval(prefix)?[0];
+        let r = verify_rand_prng.next().unwrap();
+
+        // [BBCG+21, Appendix C.4]
+        //
+        // $(z_\sigma, z^*_\sigma, z^{**}_\sigma)$
+        let tmp = r * v;
+        z[0] += tmp;
+        z[1] += r * tmp;
+        z[2] += r * k;
+        output_share.push(v);
+    }
+
+    // [BBCG+21, Appendix C.4]
+    //
+    // Add blind shares $(a_\sigma b_\sigma, c_\sigma)$
+    //
+    // NOTE(cjpatton) We can make this faster by a factor of 3 by using three seed shares instead
+    // of one.
+    let mut prng =
+        Prng::<I::Field>::from_key_stream(KeyStream::from_key(&input_share.sketch_start_seed))
+            .skip(3 * i);
+    z[0] += prng.next().unwrap();
+    z[1] += prng.next().unwrap();
+    z[2] += prng.next().unwrap();
+
+    let (d, e) = match &input_share.sketch_next {
+        Share::Leader(data) => (data[2 * i], data[2 * i + 1]),
+        Share::Helper { seed, length: _ } => {
+            // NOTE(cjpatton) We can make this faster by a factor of 2 by using two seed shares
+            // instead of one.
+            let mut prng = Prng::<I::Field>::from_key_stream(KeyStream::from_key(seed)).skip(2 * i);
+            (prng.next().unwrap(), prng.next().unwrap())
+        }
+    };
+
+    Ok((
+        VerifyState::Start { output_share, d, e },
+        VerifyMessage::Start { z },
+    ))
+}
+
+/// The VDAF verify-next algorithm for heavy hitters. The protocol has two rounds, so this is only
+/// called once by each aggregator.
+#[allow(clippy::many_single_char_names)]
+pub fn hits_next<F, M>(
+    state: VerifyState<F>,
+    msg: M,
+) -> Result<(VerifyState<F>, VerifyMessage<F>), VdafError>
+where
+    F: FieldElement,
+    M: IntoIterator<Item = VerifyMessage<F>>,
+{
+    // Compute polynomial coefficients.
+    let z = msg
+        .into_iter()
+        .map(|msg| match msg {
+            VerifyMessage::Start { z } => Ok(z),
+            _ => Err(VdafError::Uncategorized(
+                "hits_next() called on unexpected message".to_string(),
+            )),
+        })
+        .reduce(|z_left, z_right| {
+            let z_left = z_left?;
+            let z_right = z_right?;
+            Ok([
+                z_left[0] + z_right[0],
+                z_left[1] + z_right[1],
+                z_left[2] + z_right[2],
+            ])
+        })
+        .unwrap()?;
```

Comment:
nit: maybe something like
```suggestion
        .unwrap_or_else(Err(VdafError::Uncategorized("empty verify message iterator".to_string()))?;
```
IIUC the `unwrap()` handles the case where the iterator passed to `.reduce()` is empty, and `reduce()` will return `None`.